### ISSUE TITLE
FE search and sorting

### ISSUE THREAD

**Issue by datenangebot** [nextcloud/tables#201]
- [ ] text based search in FE
- [ ] sorting by clicking on column names

**Comment by jancborchardt** [comment]
UI-wise I think it would be best to combine it with the existing "Filter tables" search field on top of the left navigation.
That way it will work for the current table, but also show you other tables containing that value. This is very handy when searching for people, or specific inventory items, etc.
And people will not get confused as to which search does what – considering we also have the unified search in the top right.

**Comment by xbluemonkx** [comment]
> That way it will work for the current table, but also show you other tables containing that value.

Please keep in mind that you might want to only search a column and not a table. You might have a movide table with columns: Name | Rating.

If you are looking for movies rated ten you don't want the search to show "10 things i hate about you" because it has 10 in the title.

**Comment by xbluemonkx** [comment]
Not sure if this idea is inline with the Nextcloud CI, but could you have some sort of "toolbar" on the top that is only showing icons: Plus, Magnifier Glass. And once you hover over an icon the text next to the icon is shown?
![search](https://user-images.githubusercontent.com/28782306/232421713-6e23ab72-477a-4375-a6ec-7c6d9671c96f.PNG)

**Comment by datenangebot** [comment]
> And once you hover over an icon the text next to the icon is shown?

Interesting idea. Technically I would say yes, possible. But I guess we get some weird spaces between the icons?!

**Comment by xbluemonkx** [comment]
the space of the next needs to be dynamic.

not-hovered = 0
hovered = length of text.

In my head this happens: you hover over the + and the text slides in (left to right) and pushes the magnifier glass to the right.

probably way to much coding effort for a visual effect, that can be solved with a good static UI well.

**Comment by jancborchardt** [comment]
@xbluemonkx I really like your idea with the search via table headers, makes a lot of sense for searching in a table! :)

**Comment by datenangebot** [comment]
Lets say we have option/icons for adding new row, searching and filtering and whatever. Hover over the icons will be like a dance party with jumping (sliding) icons and texts?!

**Comment by xbluemonkx** [comment]
> Lets say we have option/icons for adding new row, searching and filtering and whatever. Hover over the icons will be like a dance party with jumping (sliding) icons and texts?!

hmmm - now that you mention that I can see the problem :/ - Probably you are right and discussing that once a prototype is ready is a better idea.

getting back to the overall topic: We just discussed the "search-whole-table"-feature. I am not even sure if this is high priority. Maybe you can postpone this and implement filter first? Filter must be column-wise. It must pe possible to filter Column 1 for "A" *AND* Filter Column 2 for "B". - Filtering a Column needs a column-wise search.

It might be, that the table-wide search might not even have a priority anymore once filtering is implemented.

**Comment by datenangebot** [comment]
> search-whole-table

This is an very interesting point: filtering vs. searching
Thanks for outlining this!

What is the most common usecase?
1. Open a table and want to find any information (text string) -> don't want to think about which columns could be effected...
2. Filter columns to get a specific view
3. Or a combination: (Idea)
Wait, I will make a blueprint, that's easier...

**Comment by datenangebot** [comment]
I would say the input field could be lighter designed.
If you insert a string we search for this everywhere,
If we type something like "What=tables app" -> enter, we get a filter displayed in the effected column...

What do you think?

<img width="703" alt="Screenshot 2023-04-17 at 11 18 31" src="https://user-images.githubusercontent.com/55329475/232441778-cb508162-50ba-4494-b0e5-4f5df145f40c.png">

**Comment by xbluemonkx** [comment]
great you brought up the use cases!

1) Search: Searching is done when you need to *find* something. You do this mostly in *unstructured* data. E.g. you have a book, long text, blog with different articles and *somewhere* there must be the thing you are searching for.
2) Filter: Filtern is done to get a subset of *structured* information. E.g. you have multiple items of the same kind and you want to see only those items that match one or more attributes.

Tables *are* *structured* information. *All* attributes of the items are known, because each attribute is a column of the table.

I am working a lot with structured tables and the only reason to *search* and not filter a table is, if I know the *value* of an attribute but not the name of it. E.g. lets say i want to filter every politicans from germany, but I do not know if the attribute is called "Country" or "State" or "Origin" etc. Then I would search for "germany" see the column (attribute) and afterwards *filter* that column for germany.

But this use case only applies to very huge tables. Looking at the Nextcloud tables App I can not imagine that someone will create a table containing 60 columns. I am most likely able to see all columns (attributes) all the time and there will be no need to *search* first to find the attribute-name.

I think the app will benefit much more if you invest the dev-time into a powerfull filter. - I recommend to look at excel filter-feature to get inspiration.

**Comment by Larsene** [comment]
why simply using datatables plugin for exemple for displaying tables ? You have all ordering, filtering, etc. tools you need !

**Comment by datenangebot** [comment]
done and merged into main, but waiting for design review
